File: gifLib/src/main/java/com/github/forjrking/gifLib/GifLibEncoder.java
package com.github.forjrking.gifLib;

import androidx.annotation.NonNull;
import android.util.Log;

import com.bumptech.glide.load.EncodeStrategy;
import com.bumptech.glide.load.Options;
import com.bumptech.glide.load.ResourceEncoder;
import com.bumptech.glide.load.engine.Resource;
import com.bumptech.glide.util.ByteBufferUtil;

import java.io.File;
import java.io.IOException;
import java.nio.ByteBuffer;

import pl.droidsonroids.gif.GifDrawable;

public class GifLibEncoder implements ResourceEncoder<GifDrawable> {

    private static final String TAG = "GifLibEncoder";

    @NonNull
    @Override
    public EncodeStrategy getEncodeStrategy(@NonNull Options options) {
        return EncodeStrategy.SOURCE;
    }

    @Override
    public boolean encode(@NonNull Resource<GifDrawable> data, @NonNull File file,
                          @NonNull Options options) {

        boolean success = false;
        if (data instanceof GifLibDrawableResource) {
            ByteBuffer byteBuffer = ((GifLibDrawableResource) data).getBuffer();
            if (byteBuffer != null) {
                try {
                    ByteBufferUtil.toFile(byteBuffer, file);
                    success = true;
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            // DES: 将 resource 编码成文件
            Log.d(TAG, String.format("GifLibEncoder -> %s -> %s", success, file.getAbsolutePath()));
        }
        return success;
    }
}

File: gifLib/src/main/java/com/github/forjrking/gifLib/GifLibEncoder.kt
package com.github.forjrking.gifLib

import android.util.Log
import com.bumptech.glide.load.EncodeStrategy
import com.bumptech.glide.load.Options
import com.bumptech.glide.load.ResourceEncoder
import com.bumptech.glide.load.engine.Resource
import com.bumptech.glide.util.ByteBufferUtil
import pl.droidsonroids.gif.GifDrawable
import java.io.File
import java.io.IOException

class GifLibEncoder : ResourceEncoder<GifDrawable?> {

    override fun getEncodeStrategy(options: Options): EncodeStrategy {
        return EncodeStrategy.SOURCE
    }

    override fun encode(data: Resource<GifDrawable?>, file: File, options: Options): Boolean {
        var success = false
        if (data is GifLibDrawableResource) {
            val byteBuffer = data.buffer
            try {
                ByteBufferUtil.toFile(byteBuffer, file)
                success = true
            } catch (e: IOException) {
                e.printStackTrace()
            }
            // DES: 将 resource 编码成文件
            Log.d(TAG, "GifLibEncoder -> $success -> ${file.absolutePath}")
        }
        return success
    }
}